File: automedal/agent/phases/researcher.py
"""Researcher phase — arxiv-driven idea injection into research_notes.md."""

from __future__ import annotations

from typing import Any

from automedal.agent.events import EventSink
from automedal.agent.kernel import RunReport
from automedal.agent.phases._common import run_phase
from automedal.agent.tools import (
    READ_FILE, WRITE_FILE, EDIT_FILE, LIST_DIR, GREP,
    RECALL, ARXIV_SEARCH, make_subagent_tool,
)


def _researcher_tools(provider, events: EventSink | None):
    base = [READ_FILE, WRITE_FILE, EDIT_FILE, LIST_DIR, GREP, RECALL, ARXIV_SEARCH]
    sub = make_subagent_tool(provider=provider, parent_tools=base, events=events)
    return base + [sub]


async def run(
    *,
    provider,
    events: EventSink | None,
    exp_id: str,
    trigger: str,
    stagnating: bool | str,
    scheduled_research: bool | int,
    best_val_loss: float | str,
    max_steps: int = 30,
) -> RunReport:
    return await run_phase(
        phase="researcher",
        provider=provider,
        tools=_researcher_tools(provider, events),
        events=events,
        max_steps=max_steps,
        slots={
            "exp_id": exp_id,
            "trigger": trigger,
            "stagnating": stagnating,
            "scheduled_research": int(bool(scheduled_research)),
            "best_val_loss": best_val_loss,
        },
    )

File: internal/agent/phases/researcher.go
// Researcher phase — arxiv-driven idea injection into research_notes.md.
package phases

import (
	"context"

	"github.com/Flameingmoy/automedal/internal/agent"
	"github.com/Flameingmoy/automedal/internal/agent/tools"
)

// ResearcherArgs is the slot dict for the researcher phase.
type ResearcherArgs struct {
	ExpID              string
	Trigger            string
	Stagnating         any // bool or string
	ScheduledResearch  any // bool/int or string
	BestValLoss        any
	MaxSteps           int
}

// RunResearcher runs the researcher phase.
func RunResearcher(ctx context.Context, chat agent.ChatStreamFunc, events *agent.EventSink, a ResearcherArgs) (agent.RunReport, error) {
	base := []tools.Tool{
		tools.ReadFile, tools.WriteFile, tools.EditFile, tools.ListDir, tools.Grep,
		tools.Recall, tools.ArxivSearch,
	}
	// Subagent inherits the base toolset.
	sub := agent.MakeSubagentTool(agent.SubagentConfig{
		Chat:        chat,
		ParentTools: base,
		Events:      events,
	})
	all := append([]tools.Tool{}, base...)
	all = append(all, sub)

	scheduled := 0
	if v, ok := a.ScheduledResearch.(bool); ok && v {
		scheduled = 1
	} else if v, ok := a.ScheduledResearch.(int); ok && v != 0 {
		scheduled = 1
	}

	maxSteps := a.MaxSteps
	if maxSteps <= 0 {
		maxSteps = 30
	}

	return RunPhase(ctx, RunPhaseConfig{
		Phase:    "researcher",
		Chat:     chat,
		Tools:    all,
		Events:   events,
		MaxSteps: maxSteps,
		Slots: map[string]any{
			"exp_id":             a.ExpID,
			"trigger":            a.Trigger,
			"stagnating":         a.Stagnating,
			"scheduled_research": scheduled,
			"best_val_loss":      a.BestValLoss,
		},
	})
}
